### PR TITLE
[v14 ready] Implement metadata for `Reactions`s

### PR DESCRIPTION
**Pull request by TorkelE**
The `Reaction` type now have a `.metadata` field:
```julia
    """
    Contain additional data, such whenever the reaction have a specific noise-scaling expression for
    the chemical Langevin equation.
    """
    metadata::Dict{Symbol, R}
```

I would have used `Base.ImmutableDict`s, but there are problems with regards to these: https://discourse.julialang.org/t/create-immutabledict-with-values-of-differenrt-types/107831

The metadata can be accessed via the following functions:
```
get_metadata_dict
has_metadata
get_metadata
```

Metadata can either be added as an optional argument when creating `Reactions`s programmatically:
```julia
using Catalyst

@variables t
@parameters k
@species X(t) X2(t)

metadata = Dict(:noise_scaling => 0.0)
r = Reaction(k, [X], [X2], [2], [1]; metadata=metadata)
```
Here, the accessors works like this:
```julia
get_metadata_dict(r)
has_metadata(r, :noise_scaling)
get_metadata(r, :noise_scaling)
```

Metadata can also be provided via the DSLs:
```julia
rs = @reaction_network begin
    @parameters η
    k, 2X --> X2, [noise_scaling=η]
    k, 2X --> X2, [some_metadata=1.0, more_metadata="Hello world"]
end

@parameters η
@reaction k, 2X --> X2, [noise_scaling=$η]
@reaction k, 2X --> X2, [some_metadata=1.0, more_metadata="Hello world"]
```

Finally, instead of using arrows like `=>`, 
```julia
rs = @reaction_network begin
    k, 2X --> X2, [only_use_rate=true]
end
```
can be used to remove the substrate expression from the rate. Internally, `only_use_rate=true` is carried through as a metadata to record this. However, it is removed at the end.

Currently, no additional metadata is actually used. However, `noise_scaling` will be added after this is merged. 

Also:
- I have updated this, adding `Bool`: `const ExprValues = Union{Expr, Symbol, Float64, Int, Bool}`. Here, `Bool`s can potentially be inside expressions (like `Float64`). We missed this before and I realised when working with this. Added it now.
- Since some tests here are specific to the `Reaction` struct, I added another test file for tests related to this struct.
- I made a minor internal remake so that reactions created in `@reaction` and `@reaction_network` are created using the same internal function.

Considerations:
- Should `only_use_rate` remain in the metadata post-creation? This could also be entirely turned into a metadata thing. Alternatively, we keep it as currently implemented, where this can be used as an option, but never actually appear in the metadata.
- Should we maintain a list of permitted metadata? This would prevent the user to add their own, but would give us more control. Alternatively, we can add a warning if another metadata was used. This would be useful if someone e.g. writers `noise_scalling = `, since we could then catch this error.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
It probably doesn't make too much of a difference at this level, but I think the continuity of having it match Symbolics and ModelingToolkit's metadata system would make things cleaner.

**Comment by isaacsas** [comment]
I don’t think we should use Dicts though. That seems pretty heavy for when we have systems with thousands of reactions. Looking at that thread I’d say we are in the use case for ImmutableDicts (small numbers of objects being stored). And we should settle the interface now, otherwise it will end up a breaking change in the future.

**Comment by isaacsas** [comment]
Alternatively we could just use a flat vector and linear search given we don’t expect lots of metadata.

**Comment by TorkelE** [comment]
If we go down the MTK route, we are still happy with the other parts, especially how `only_use_rate` is handled?

Is there a good resource on the MTK thing? I don't dislike the idea of using that interface, but I don't want to get bogged down and spend the next couple of days looking at a very niche implementation that we still rate not sure will actually matter within the next couple of years (especially since both I and Catalyst have better stuff to concentrate on).

Basically, if we want to go down the MTK route I would want to know what we want and how now, rather than learning details over 3 different revisions of a PR.

**Comment by TorkelE** [comment]
Another alternative should be to simply make it a named Tuple? That would allow indexing of the fields, make it immutable, and be good for storage? That should work equivalently and also be relatively light weigth.

**Comment by TorkelE** [comment]
The last line here is what causes the error:
```julia
@variables t
@parameters k
@species X(t) X2(t)

metadata = [:md_1 => 1.0, :md_2 => false, :md_3 => "Hello world", :md_4 => :sym, :md_5 => X + X2^k -1, :md_6 => (0.1, 2.0)]
```
No idea why, it all works fine locally.

**Comment by TorkelE** [comment]
Ok, this one is ready now.

**Review by isaacsas — CHANGES_REQUESTED** [review]
@TorkelE can you remind me why this has diverged so far from the MTK metadata approach / interface, i.e. not declaring new metadata as in https://github.com/SciML/Catalyst.jl/blob/faa8fc744168f1422639fb7f77c900439fd6171c/src/reactionsystem.jl#L1-L7? It seems confusing to users to have one way to declare new metadata for symbolic variables and a completely different symbol based approach for reactions. Why not just reuse the MTK notion of what a metadata indicator is?

File: docs/src/api/catalyst_api.md [hunk 1]
```diff
@@ -178,6 +178,9 @@ substoichmat
 prodstoichmat
 netstoichmat
 reactionrates
+get_metadata_dict
```
Comment (isaacsas) [inline docs/src/api/catalyst_api.md:181]: This seems like a different approach than what we use for symbolics. There we define custom get/has functions for each possible metadata type, not generic functions that take a symbol indicating the metadata we want. Why change the interface here?

**Comment by isaacsas** [comment]
The reason for Symbolic's interface doesn't really matter here though -- what matters is that now we are using a completely different interface. (I'm not talking about the internal implementation details but the user facing way to add new metadata, query its existence as an option, and get its value when present in a given `Reaction`.) However we internally implement the details why not make the external interface for declaring a new metadata type, querying existence of a given type within a reaction, and getting values of a specific type of metadata the same?

**Comment by TorkelE** [comment]
On the airport right now, but will think more on it when I get back.

While I see the advantage of having the same interface, this one is just insanely much easier to use. The Symbolics one feels really internal, and is nothing I see anyone using outside of package development.

If you want we can simply remove the doc part of this one, and let it be internal for now though.

**Comment by TorkelE** [comment]
Letäs discuss this one in person, I am still a bit uncertain exactly what you envision would be removed/added/external/internal.

**Comment by TorkelE** [comment]
This is now ready @isaacsas

File: src/reactionsystem.jl [hunk 2]
```diff
@@ -181,7 +186,27 @@ function Reaction(rate, subs, prods, substoich, prodstoich;
...
+    # Ensures metadata have the correct type.
+    metadata = convert(Vector{Pair{Symbol, Any}}, metadata)
+
+    Reaction(value(rate), subs, prods, substoich′, prodstoich′, ns, only_use_rate, metadata)
+end
+
+# Checks if a metadata input has an entry :only_use_rate => true
+function metadata_only_use_rate_check(metadata)
+    any(:only_use_rate == entry[1] for entry in metadata) || (return false)
+    return Bool(metadata[findfirst(:only_use_rate == entry[1] for entry in metadata)][2])
```
Comment (isaacsas) [inline src/reactionsystem.jl:209]: Why not just use `findfirst` and if it returns `nothing` you know that it isn't defined? You shouldn't need two searches through the `metadata`.

**Comment by isaacsas** [comment]
@TorkelE once these are addressed we can merge.

File: src/Catalyst.jl [hunk 3]
```diff
@@ -67,7 +67,7 @@ export ODEProblem,
        SteadyStateProblem
 
 # reaction_network macro
-const ExprValues = Union{Expr, Symbol, Float64, Int}
+const ExprValues = Union{Expr, Symbol, Float64, Int, Bool}
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply src/Catalyst.jl:70]: Maybe when all this is over we check whether changing makes a difference, and if better, we just change this to `Any`.

File: HISTORY.md [hunk 4]
```diff
@@ -16,6 +16,15 @@ assess_identifiability(goodwind_oscillator; measured_quantities=[:M])
...
 - Adds a tutorial to illustrate the use of the extension.
+- Enable adding metadata to individual reactions, e.g:
+```julia
+rn = @reaction_network begin
+    @parameters η
+    k, 2X --> X2, [noise_scaling=η]
+end
+getnoisescaling(rn)
+```
+- Changed fields of internal `Reaction` structure. `ReactionSystems`s saved using `serialize` on previous Catalyst versions cannot be loaded using this (or later) versions.
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply HISTORY.md:27]: I can just remove this line then?

File: src/reaction_network.jl [hunk 5]
```diff
@@ -580,70 +581,88 @@ end
...
-                             get_tup_arg(rate, i), only_use_rate))
+# Takes a reaction line and creates reaction(s) from it and pushes those to the reaction array.
+# Used to create multiple reactions from, for instance, `k, (X,Y) --> 0`.
+# Handles metadata, e.g. `1.0, Z --> 0, [noisescaling=η]`.
+function push_reactions!(reactions::Vector{ReactionStruct}, sub_line::ExprValues, prod_line::ExprValues, 
+                         rate::ExprValues, metadata::ExprValues, arrow::Symbol)  
+    # The rates, substrates, products, and metadata may be in a tupple form (e.g. `k, (X,Y) --> 0`).
+    # This finds the length of these tuples (or 1 if not in tuple forms). Errors if lengs inconsistent.
+    lengs = (tup_leng(sub_line), tup_leng(prod_line), tup_leng(rate), tup_leng(metadata))
+    if any(!(leng in [1, maximum(lengs)]) for leng in lengs)
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply src/reaction_network.jl:646]: goo catch, agreed

File: src/reaction_network.jl [hunk 6]
```diff
@@ -580,70 +581,88 @@ end
...
+        throw("Malformed reaction, rate=$rate, subs=$sub_line, prods=$prod_line, metadata=$metadata.")
+    end
+
+    # Loops through each reaction encoded by the reaction composites. Adds the reaction to the reactions vector.
+    for i in 1:maximum(lengs)                       
+        # If the `only_use_rate` metadata was not provided, this has to be infered from the arrow used.
+        metadata_i = get_tup_arg(metadata, i)
+        if all([arg.args[1] != :only_use_rate for arg in metadata_i.args])
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply src/reaction_network.jl:654]: it should

File: src/reaction_network.jl [hunk 7]
```diff
@@ -580,70 +581,88 @@ end
...
+            push!(metadata_i.args, :(only_use_rate = $(in(arrow, pure_rate_arrows))))
+        end
+
+        # Checks that metadata fields are unqiue.
+        metadata_entries = [arg.args[1] for arg in metadata_i.args]
+        if length(unique(metadata_entries)) < length(metadata_entries)
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply src/reaction_network.jl:660]: Didn't know about `allunique`, that is useful!

File: HISTORY.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (isaacsas) [reply HISTORY.md:27]: We can have it here, but it probably needs to be right upfront in the full v14 release comments and not buried.

File: src/reactionsystem.jl [hunk 8]
```diff
@@ -181,7 +186,27 @@ function Reaction(rate, subs, prods, substoich, prodstoich;
         convert.(stoich_type, netstoich) : netstoich
     end
 
-    Reaction(value(rate), subs, prods, substoich′, prodstoich′, ns, only_use_rate)
+    # Check that all metadata entries are unique. (cannot use `in` since some entries may be symbolics).
+    if any(any(isequal(metadata[i][1], entry[1]) for entry in metadata[i+1:end]) 
+                                                 for i in eachindex(metadata))
+        error("Repeated entries for the same metadata encountered in the following metadata set: $([entry[1] for entry in metadata]).")
+    end
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply src/reactionsystem.jl:193]: `allunique` works for Symbolics, changed to using it.

File: src/reactionsystem.jl (hunk 2 above)
Comment (TorkelE) [reply src/reactionsystem.jl:209]: yes, that is better, fixed.

File: src/reactionsystem.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (isaacsas) [reply src/reactionsystem.jl]: That is because it is based on hashing, which works on Symbolic stuff (as does `isequal`, hence why it works).

[the remark this answers is not in the document]
Comment (TorkelE) [reply src/reactionsystem.jl]: That sounds like a good way to do it. Also like a (potentially?) O(n) time to check, without allocating (which I couldn't directly think of).
